### PR TITLE
refactor ZoneAwareClusterInvoker.

### PR DESCRIPTION
**Pull request by zonghaishang**
## What is the purpose of the change

XXXXX

## Brief changelog

XXXXX

## Verifying this change

XXXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Run `mvn clean install -DskipTests=false` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/registry/ZoneAwareClusterInvoker.java [hunk 1]
```diff
@@ -148,60 +146,55 @@ public Result doInvoke(Invocation invocation, final List<Invoker<T>> invokers, L
...
-                    }
-                } else {
-                    if (logger.isDebugEnabled()) {
-                        logger.debug("step is APPLICATION_FIRST serviceInvokers is empty get interfaceInvokers");
                     }
-                    return interfaceInvokers;
                 }
 
+                if (logger.isDebugEnabled()) {
+                    logger.debug("step is APPLICATION_FIRST serviceInvokers " + ((!serviceAvailable)? "is empty": "shouldMigrate false") + " get interfaceInvokers");
```
Comment (horizonzy) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/registry/ZoneAwareClusterInvoker.java:190]: `serviceInvokers` should be judge by `serviceAvailable`.The logic isn't same as before.

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/registry/ZoneAwareClusterInvoker.java [hunk 2]
```diff
@@ -216,23 +209,16 @@ public Result doInvoke(Invocation invocation, final List<Invoker<T>> invokers, L
 
     private boolean shouldMigrate(boolean addressChanged, List<Invoker<T>>  serviceInvokers, List<Invoker<T>>  interfaceInvokers) {
```
Comment (horizonzy) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/registry/ZoneAwareClusterInvoker.java:210]: The indent can be format followed the checkstyle

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/registry/ZoneAwareClusterInvoker.java [hunk 3]
```diff
@@ -216,23 +209,16 @@ public Result doInvoke(Invocation invocation, final List<Invoker<T>> invokers, L
...
-                return  false;
-            } else {
-                return  true;
-            }
+        if (detectors != null && !detectors.isEmpty()) {
+            return detectors.stream().allMatch(s -> s.shouldMigrate(interfaceInvokers, serviceInvokers));
         }
+
+        // check application level provider available.
+        List<Invoker<T>>  availableServiceInvokers = serviceInvokers.stream().filter( s -> s.isAvailable()).collect(Collectors.toList());
```
Comment (horizonzy) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/registry/ZoneAwareClusterInvoker.java:217]: indent can be format followed the checkstyle.

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/registry/ZoneAwareClusterInvoker.java [hunk 4]
```diff
@@ -148,60 +146,55 @@ public Result doInvoke(Invocation invocation, final List<Invoker<T>> invokers, L
...
-            } else {
-                // 不一致
-                if (!rule.equals(migrationClusterInvoker.getMigrationRule())) {
-                    rule = MigrationRule.queryRule();
-                    break;
-                }
+                continue;
+            }
+
+            // 不一致
```
Comment (horizonzy) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/registry/ZoneAwareClusterInvoker.java:157]: in english.

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/registry/ZoneAwareClusterInvoker.java (hunk 1 above)
Comment (zonghaishang) [reply dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/registry/ZoneAwareClusterInvoker.java:190]: `serviceInvokers.size() > 0` equals `!serviceInvokers.isEmpty()` 
`serviceAvailable` equals `serviceInvokers.size() > 0`

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/registry/ZoneAwareClusterInvoker.java (hunk 2 above)
Comment (zonghaishang) [reply dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/registry/ZoneAwareClusterInvoker.java:210]: fixed it.

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/registry/ZoneAwareClusterInvoker.java (hunk 4 above)
Comment (zonghaishang) [reply dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/registry/ZoneAwareClusterInvoker.java:157]: fixed it.

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/registry/ZoneAwareClusterInvoker.java (hunk 3 above)
Comment (zonghaishang) [reply dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/registry/ZoneAwareClusterInvoker.java:217]: fixed it.

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/registry/ZoneAwareClusterInvoker.java [hunk 5]
```diff
@@ -148,60 +146,55 @@ public Result doInvoke(Invocation invocation, final List<Invoker<T>> invokers, L
...
-                    }
-                } else {
-                    if (logger.isDebugEnabled()) {
-                        logger.debug("step is APPLICATION_FIRST serviceInvokers is empty get interfaceInvokers");
                     }
-                    return interfaceInvokers;
                 }
 
+                if (logger.isDebugEnabled()) {
+                    logger.debug("step is APPLICATION_FIRST serviceInvokers " + (serviceInvokers.isEmpty() ? "is empty" : "shouldMigrate false") + " get interfaceInvokers");
```
Comment (horizonzy) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/registry/ZoneAwareClusterInvoker.java:190]: 额。我的意思是把serviceInvokers这个str移到三元表达式里面。
logger.debug("step is APPLICATION_FIRST " + (serviceInvokers.isEmpty() ? "serviceInvokers is empty" : "shouldMigrate false") + " get interfaceInvokers");

Comment (zonghaishang) [reply dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/registry/ZoneAwareClusterInvoker.java:190]: fixed it.

**Review by horizonzy — APPROVED** [review]
LGTM.

**Review by AlbumenJ — APPROVED** [review]
LGTM.

**Pull request merged**